tungsten_ci_utils/nexus_cleanup/clean_up.groovy

```diff
@@ -0,0 +1,97 @@
+import org.sonatype.nexus.repository.storage.StorageFacet;
+import org.sonatype.nexus.common.app.GlobalComponentLookupHelper;
+import org.sonatype.nexus.repository.maintenance.MaintenanceService;
+import org.sonatype.nexus.repository.storage.ComponentMaintenance;
+import org.sonatype.nexus.repository.storage.Query;
+import org.sonatype.nexus.script.plugin.RepositoryApi;
+import org.sonatype.nexus.script.plugin.internal.provisioning.RepositoryApiImpl;
+import com.google.common.collect.ImmutableList;
+import org.joda.time.DateTime;
+import org.slf4j.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+def retentionDays = 15;
+def retentionBorder = 15;
+def repositoryName = 'BartsDockerRepo';
+def whitelisted_tag_suffixes = ["ocata", "newton", "latest", "40", "94", "122", "129", "161", "214", "309", "360"].toArray();
+def retentionCounter = 0;
+log.info(":::Cleanup script started!");
+MaintenanceService service = container.lookup("org.sonatype.nexus.repository.maintenance.MaintenanceService");
+def repo = repository.repositoryManager.get(repositoryName);
+def tx = repo.facet(StorageFacet.class).txSupplier().get();
+def components = null;
+try {
+    tx.begin();
+    components = tx.browseComponents(tx.findBucket(repo));
+} catch (Exception e) {
+    log.info("Error: " + e);
+} finally {
+    if (tx != null)
+        tx.close();
+}
+
+if (components != null) {
+    def retentionDate = DateTime.now().minusDays(retentionDays).dayOfMonth().roundFloorCopy();
+    int deletedComponentCount = 0;
+    int compCount = 0;
+    def listOfComponents = ImmutableList.copyOf(components);
+    def previousComp = listOfComponents.head().name();
+    def checkValue = null;
+    retentionCounter = 0;
+    def prev = null;
+    listOfComponents.reverseEach {  comp ->
+
+        def splited = comp.version();
+        def spl = splited.split("-");
+
+        for (i = spl.length - 1; i < spl.length; i++) {
+            checkValue = null;
+            for (j = 0; j < whitelisted_tag_suffixes.length; j++) {
+                Pattern pattern = Pattern.compile(whitelisted_tag_suffixes[j]);
+                Matcher matcher = pattern.matcher(spl[i]);
+                boolean found = matcher.matches();
+                if (found == true) {
+                    log.info("true " + j);
+                    checkValue = true;
+                    log.info("Component skipped: ${comp.name()} ${comp.version()}");
+                    return checkValue;
+                }
+            }
+            log.info("else check " + prev + " " + spl[i - 1]);
+            if ((spl[i - 1]) == prev) {
```

Comment:
Dont think this logic with prev makes sense. This would only leave the last 15 tags for e.g. build number 290 so the case matches both queens-master-290 and rhel-queens-master-290 when they should be considered separately.

Reply:
Additionally I think the tx.browseComponents(tx.findBucket(repo)) returns all images and their tags. You need to make sure that each particular image repo is handled separately (that is: leave 15 latest tags and not older than 15 days for contrail-vrouter-agent and contrail-nodemgr etc.). Heres a list of all docker images and their tags on dockerhub: https://hub.docker.com/r/opencontrailnightly/